### ISSUE TITLE
FR: Add a `jj track` command for tracking otherwise ignored files

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#2837]
**Is your feature request related to a problem? Please describe.**

If you have ignored files that you want to track, you currently have to temporary un-ignore them, which can be quite annoying, especially if they're in a directory that's ignored (unless you want to track everything in that directory).

**Describe the solution you'd like**

`jj track <pattern>`  seems logical (we already have `jj untrack <pattern>`).

**Comment by benbrittain** [comment]
`track` is a pretty easy command to implement, but it has some interesting overlap with the forget mechanism. If there is a `track`, should `untrack` automatically add a file to be forgotten?

**Comment by martinvonz** [comment]
`untrack` doesn't currently automatically add files to the ignore patterns, mostly because it's not clear if the user wants that particular file or a more general pattern to be ignored. @hooper, did you say you're also looking into this?

**Comment by hooper** [comment]
Not really.

I'm trying to implement `track` by refactoring `GitIgnoreFile` into a trait to allow it to interact with an impl that uses a `Matcher`, so I can just construct that and plug it into `snapshot()`.

It would be a more invasive change to act further on the decisions to ignore paths during snapshot by persisting them. I agree with Martin that there are issues with defining a good UX for that.

Such a feature could become unpopular for spamming `.gitignore` files with confusing entries, even if we do some basic optimizations. Then it might turn into a tempting target for scope creep to "fix" that. Also, it's not obvious to me that every `<pattern>` we might want to accept would map reasonably to `.gitignore` patterns. For examples of how that could get harder or more ambiguous in the future, see https://github.com/martinvonz/jj/issues/2915.

I've occasionally imagined something like `jj [un]ignore <pattern>` that does this more directly, but it would have at least as many issues as doing it during `[un]track`.

Another detail we would want to consider is the UX for how we determine *which* `.gitignore` file we manipulate under which circumstances. I don't think something along the lines of `jj config set --user|--repo` would be sufficient.

So, I don't mean to shoot the idea down entirely, but I don't think it's low hanging fruit.

**Comment by Nemo157** [comment]
I want this for the case where I've ignored a file in my global `~/.git/config/ignore` but I want to have it tracked in one specific repository. It's not ignored by that repository's local `.gitignore` so I don't think it's appropriate to add my own `!...` pattern to that file because of my global ignores. I did find a workaround of temporarily adding a negative pattern to `.gitignore`, running `jj st` so it starts tracking it, then deleting the pattern from `.gitignore`.

**Comment by ilyagr** [comment]
To expand on the link from #4338, the newly current state of this is that there is a `jj track` command, but it does *not* currently allow tracking otherwise ignored files. 

It should, and that is probably what resolving this issue would now entail.

See also https://github.com/martinvonz/jj/pull/4338#discussion_r1740336563

**Comment by martinvonz** [comment]
I would also consider it a blocker for closing this issue to teach `jj status` to print the untracked files. I should do that to the best of its ability anyway; note that e.g. using watchman may mean that we won't be able to list all untracked files.

**Comment by ilyagr** [comment]
I was thinking we'd have two categories of untracked files:

1. The ones that are considered explicitly ignored, whether this comes from a gitignore, from a `jj untrack`, or from being outside a sparse pattern (there might be other possibilities I forgot).
2. The ones that are not ignored but `jj` thinks the user might not actually want to snapshot. `jj` would want an explicit user confirmation before snapshotting such files. These would be files >1MB (the current behavior of `jj` with these files is clearly imperfect; what I'm talking about here might not be a sufficient solution to the problem but seems like it could be a part of it); perhaps there are other files that fit into this category like cache files (#4409) or files with secrets.

My first thought was that we only really need to show the second category in `jj status`. Is this similar to what you are thinking? What files do you want to be shown in `jj status`?

This was also discussed in https://github.com/martinvonz/jj/pull/4338#discussion_r1735046554, though I don't think there was a clear conclusion.

P.S. For the sake of completeness, there is also the crazy category of files that might be untracked, but might exist in both the working copy and the working copy commit (and might differ!). So far, I think we can consider the existence of such files as an error one way or another. This is in spite of the fact that this category is fine for Git, it only seems crazy in `jj`-land.

**Comment by ilyagr** [comment]
Of course it would be good to have a way to see which of the physical files in the working copy are being tracked (and, ideally, to debug why), but I'm not sure whether that's what you meant or why you think it's important or whether you think it should be the default.

**Comment by martinvonz** [comment]
> My first thought was that we only really need to show the second category in `jj status`. Is this similar to what you are thinking? What files do you want to be shown in `jj status`?

Yes, untracked-but-not-ignored files is exactly what I want `jj status` to show.

> For the sake of completeness, there is also the crazy category of files that might be untracked, but might exist in both the working copy and the working copy commit (and might differ!).

We don't currently support that and I don't think it seems worth the complexity to support (but since I'm not motivated to support that case, I have spent basically no time trying to figure out how hard it would be).

**Comment by martinvonz** [comment]
> Of course it would be good to have a way to see which of the physical files in the working copy are being tracked (and, ideally, to debug why or why not)

We already have `jj file list` for that (minus the debugging part), right?

> but I'm not sure whether that's what you meant or why you think it's important or whether you think it should be the default.

I think it's important for people who want to use `snapshot.auto-track = "none()"` to be able to easily figure out which paths to pass to `jj track`. They currently have to use `git status` for that, which is obviously not ideal.

**Comment by ilyagr** [comment]
> Yes, untracked-but-not-ignored files is exactly what I want jj status to show.

I think that in this comment, you're talking about a different thing than I was.

> I think it's important for people who want to use snapshot.auto-track = "none()" to be able to easily figure out which paths to pass to jj track. They currently have to use git status for that, which is obviously not ideal.

This makes more sense to me regarding what you mean by "untracked-but-not-ignored" files. I haven't yet thought about it much, but it's a usecase that makes sense, and I wanted to post a quick comment to make sure we're on the same page before thinking about it.

> > For the sake of completeness, there is also the crazy category of files that might be untracked, but might exist in both the working copy and the working copy commit (and might differ!).
>
> We don't currently support that and I don't think it seems worth the complexity to support it (but since I'm not motivated to support that case, I have spent basically no time trying to figure out how hard it would be).

+1, though I have not thought about it beyond what I wrote down either.

**Comment by joyously** [comment]
I still think there shouldn't be multiple sources of truth. I think `untrack` and `track` should simply open the editor with the ignore file. It should be an atomic operation.

**Comment by stacyharper** [comment]
Maybe we should update the documentation that currently says:

```
Ignored files are never tracked automatically (regardless of the value
of snapshot.auto-track), but they can still end up being tracked for a
few reasons:

- because of an explicit jj file track command
```

https://jj-vcs.github.io/jj/latest/working-copy/#ignored-files

**Comment by martinvonz** [comment]
Good catch. I don't know why I wrote it that way.  I sent #5902 to fix it.

**Comment by ErichDonGubler** [comment]
As an alternative to [this comment's](https://github.com/jj-vcs/jj/issues/2837#issuecomment-2674585797) proposal to add `--force <fileset>` to `jj file track`, I'd like to throw my hat in the ring by proposing a `--include-ignored` flag for `jj file track` instead. I feel like making the invocation more consistent (keeping positional args as the filesets used) and having users issue separation invocations between ignored and non-ignored files is easier UX to grok.

**Comment by ErichDonGubler** [comment]
(Should this issue be retitled `s/jj track/jj file track/`, BTW?)

**Comment by mantljosh** [comment]
I have some occasional files that are larger than the max file size that prevent them from being tracked automatically, which is okay default behavior but it is a pain to explicitly track the file since there's no `--force`. Every time this happens I have to first run `jj file track <path>` so that it will complain and then copy/paste the `jj --config snapshot.max-new-file-size=3027899 file track` command since you have to set that specifically

IMO if I'm running `jj file track` I've already expressed an intent to include this file even if it is currently ignored by the auto-tracking logic. I think a warning that "the file you just added is larger than the configured size" is reasonable but I don't think it should prevent the file from being tracked.

**Comment by PhilipMetzger** [comment]
@tingerrr had a similar idea a while ago in the Discord. I think he should add his idea as a comment here.

**Comment by tingerrr** [comment]
Yeah, I was thinking the exact same thing. If it's not tracked because of being too large, then calling `jj file track` should circumvent the snapshot check.

**Comment by mantljosh** [comment]
I've been keeping an eye out for an opportunity to jump into the JJ codebase, so if there's an appetite for a change here I'd be willing to make an attempt at putting up a PR.

**Comment by PhilipMetzger** [comment]
I mean you should feel free to take a shot at it since @hooper hasn't actively worked on this for a while now.

**Comment by martinvonz** [comment]
@mantljosh: Thanks! For the`jj file track --include-ignored` feature, I think we can do it at least two different ways:

1. Pass in a second `Matcher` into `LockedWorkingCopy::snapshot()` and use that for deciding which paths to track even if they're ignored.
2. Make the command itself (as opposed to `local_working_copy.rs`) walk the specified paths and add the paths to the tree. Then we can call `LockedWorkingCopy::reset()` to make them tracked, followed by `LockedWorkingCopy::snapshot()` to snapshot the contents.

I think either approach would also work for tracking large files. I think I like the second approach better because it doesn't involve making the `LockedWorkingCopy` trait more complicated. That means the new feature will work with Google's custom working copy implementation without modification, for example. On the other hand, the second approach means that any future command that want to implement similar behavior will have to write its own similar logic. I think that's less likely, so it seems fine to me.

**Comment by LucioFranco** [comment]
Since there has not been a response in a few weeks, I took a stab at it with #7900

**Comment by LucioFranco** [comment]
With #7900 merged, we can close this!
